### PR TITLE
JP-3347: Improve spectral outlier detection

### PR DESCRIPTION
**Pull request by melanieclarke**
Resolves [JP-3347](https://jira.stsci.edu/browse/JP-3347)

Modify outlier detection for spectral data, in order to threshold outliers for an exposure against an error value computed across exposures, instead of the input error for the exposure itself. This is intended to help address extreme outliers that have both very high data values and very high error values in the input exposure.

The process for resampled data is:
- drizzle the error array the same way as the science, to approximate a resampled error for each exposure
- median combine the drizzled error arrays across exposures, in the same way the drizzled science arrays are median combined
- blot the median error back to the input data coordinates, in the same way the median science image is handled
- compare (input data - blotted median) to blotted median error for thresholding purposes.

For outlier detection on spectra without resampling (not commonly used), the median error is directly computed across the input exposures and used for thresholding.

## Tasks
- [x] **request a review from someone specific**, to avoid making the maintainers review every PR
- [x] add a build milestone, i.e. `Build 11.3` (use the [latest build](https://github.com/spacetelescope/jwst/milestones) if not sure)
- [x] Does this PR change user-facing code / API? (if not, label with `no-changelog-entry-needed`)
  - [x] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see below for change types) 
  - [x] update or add relevant tests
  - [x] update relevant docstrings and / or `docs/` page
  - [x] [start a regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) and include a link to the running job ([click here for instructions](https://github.com/spacetelescope/RegressionTests/blob/main/docs/running_regression_tests.md))
    - [x] Do truth files need to be updated ("okified")?
      - [ ] **after the reviewer has approved these changes**, run `okify_regtests` to update the truth files
- [ ] if a JIRA ticket exists, [make sure it is resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

<details><summary>news fragment change types...</summary>

- ``changes/<PR#>.general.rst``: infrastructure or miscellaneous change
- ``changes/<PR#>.docs.rst``
- ``changes/<PR#>.stpipe.rst``
- ``changes/<PR#>.datamodels.rst``
- ``changes/<PR#>.scripts.rst``
- ``changes/<PR#>.fits_generator.rst``
- ``changes/<PR#>.set_telescope_pointing.rst``
- ``changes/<PR#>.pipeline.rst``

## stage 1
- ``changes/<PR#>.group_scale.rst``
- ``changes/<PR#>.dq_init.rst``
- ``changes/<PR#>.emicorr.rst``
- ``changes/<PR#>.saturation.rst``
- ``changes/<PR#>.ipc.rst``
- ``changes/<PR#>.firstframe.rst``
- ``changes/<PR#>.lastframe.rst``
- ``changes/<PR#>.reset.rst``
- ``changes/<PR#>.superbias.rst``
- ``changes/<PR#>.refpix.rst``
- ``changes/<PR#>.linearity.rst``
- ``changes/<PR#>.rscd.rst``
- ``changes/<PR#>.persistence.rst``
- ``changes/<PR#>.dark_current.rst``
- ``changes/<PR#>.charge_migration.rst``
- ``changes/<PR#>.jump.rst``
- ``changes/<PR#>.clean_flicker_noise.rst``
- ``changes/<PR#>.ramp_fitting.rst``
- ``changes/<PR#>.gain_scale.rst``

## stage 2
- ``changes/<PR#>.assign_wcs.rst``
- ``changes/<PR#>.badpix_selfcal.rst``
- ``changes/<PR#>.msaflagopen.rst``
- ``changes/<PR#>.nsclean.rst``
- ``changes/<PR#>.imprint.rst``
- ``changes/<PR#>.background.rst``
- ``changes/<PR#>.extract_2d.rst``
- ``changes/<PR#>.master_background.rst``
- ``changes/<PR#>.wavecorr.rst``
- ``changes/<PR#>.srctype.rst``
- ``changes/<PR#>.straylight.rst``
- ``changes/<PR#>.wfss_contam.rst``
- ``changes/<PR#>.flatfield.rst``
- ``changes/<PR#>.fringe.rst``
- ``changes/<PR#>.pathloss.rst``
- ``changes/<PR#>.barshadow.rst``
- ``changes/<PR#>.photom.rst``
- ``changes/<PR#>.pixel_replace.rst``
- ``changes/<PR#>.resample_spec.rst``
- ``changes/<PR#>.residual_fringe.rst``
- ``changes/<PR#>.cube_build.rst``
- ``changes/<PR#>.extract_1d.rst``
- ``changes/<PR#>.resample.rst``

## stage 3
- ``changes/<PR#>.assign_mtwcs.rst``
- ``changes/<PR#>.mrs_imatch.rst``
- ``changes/<PR#>.tweakreg.rst``
- ``changes/<PR#>.skymatch.rst``
- ``changes/<PR#>.exp_to_source.rst``
- ``changes/<PR#>.outlier_detection.rst``
- ``changes/<PR#>.tso_photometry.rst``
- ``changes/<PR#>.stack_refs.rst``
- ``changes/<PR#>.align_refs.rst``
- ``changes/<PR#>.klip.rst``
- ``changes/<PR#>.spectral_leak.rst``
- ``changes/<PR#>.source_catalog.rst``
- ``changes/<PR#>.combine_1d.rst``
- ``changes/<PR#>.ami.rst``

## other
- ``changes/<PR#>.wfs_combine.rst``
- ``changes/<PR#>.white_light.rst``
- ``changes/<PR#>.cube_skymatch.rst``
- ``changes/<PR#>.engdb_tools.rst``
- ``changes/<PR#>.guider_cds.rst``
</details>

### CODE REVIEW

**Comment by melanieclarke** [comment]
Regression tests show minor differences for slit spectra in spec3 tests (NIRSpec FS and MOS, MIRI LRS), downstream of outlier detection, as expected.  The MIRI differences are not significant - there are a few pixels slightly different, near a saturated region of the spectrum.  For the NIRSpec data, some new outliers are appropriately flagged and good signal does not appear to be newly flagged, with one exception.  

The spectrum in this test:
[jwst.regtest.test_nirspec_fs_spec3_moving_target.[stable-deps] test_nirspec_fs_spec3_moving_target[x1d]](https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1763/testReport/junit/jwst.regtest/test_nirspec_fs_spec3_moving_target/_stable_deps__test_nirspec_fs_spec3_moving_target_x1d_/)
shows some new over-aggressive flagging at the core of the trace.  This is similar to issues described in JP-3347, and can be fixed with the recommended parameter override scale="3.0 2.0".  The NIRSpec team plans to follow up this PR with new default parameters, so I think this is not a concern for these changes.

**Comment by melanieclarke** [comment]
I will leave this PR in draft for now, because I'd like to hear back from the MIRI team to make sure they are okay with these changes, but I think this is otherwise ready for review.

File: docs/jwst/outlier_detection/outlier_detection_spec.rst
Comment (emolter) [inline docs/jwst/outlier_detection/outlier_detection_spec.rst:39]:
```suggestion
   - The median datamodel, with ``data`` and ``err`` extensions, is written out to disk with suffix "median"
```

File: jwst/outlier_detection/outlier_detection_step.py
Comment (emolter) [inline jwst/outlier_detection/outlier_detection_step.py:54]: Is this directly related to the spectral outlier detection or related to a different issue?  (I'm fine including it either way, but just would like to know what prompted this specifically)

File: jwst/outlier_detection/utils.py
Comment (emolter) [inline jwst/outlier_detection/utils.py:71]: Random extraneous comment, but can we make the line length limit 110 characters instead of 80 when we implement the ruff check?  stcal has 110 characters, and personally I find 80 to sometimes detract from readability

File: jwst/outlier_detection/utils.py
Comment (emolter) [inline jwst/outlier_detection/utils.py:163]:
```suggestion
    The median is performed across resampled groups, for both imaging
```

File: jwst/outlier_detection/utils.py
Comment (emolter) [inline jwst/outlier_detection/utils.py:167]:
```suggestion
```

File: jwst/outlier_detection/utils.py
Comment (emolter) [inline jwst/outlier_detection/utils.py:193]:
```suggestion
        If True, an approximate median error is computed alongside the
```

File: jwst/outlier_detection/utils.py
Comment (emolter) [inline jwst/outlier_detection/utils.py:92]:
```suggestion
        If True, an approximate median error is computed alongside the
```

File: jwst/outlier_detection/utils.py
Comment (emolter) [inline jwst/outlier_detection/utils.py:296]: same question as above about setting fillval to NaN; does this enforce NaN setting related to a different issue?

File: jwst/resample/resample.py
Comment (emolter) [inline jwst/resample/resample.py:265]:
```suggestion
            If True, an approximate error image will be resampled
```

**Review by emolter — COMMENTED** [review]
I couldn't find anything wrong with the code, looks excellent to me.  But I suppose the behavior almost exactly mimics what is done with the data.  Most of my comments are minor and docstring-related.  I'll leave it to the INS teams to decide if the regtest outputs look good or not

File: jwst/outlier_detection/outlier_detection_step.py
Comment (melanieclarke) [reply jwst/outlier_detection/outlier_detection_step.py:54]: It is related - with some proposed parameter combinations, leaving fillval = INDEF does bad things to spectra, since it may combine zeros with real data.  NaN values are always ignored.

But also, the default for resampling was changed in #8488, and the outlier detection default should be changed to match, for safer and more consistent behavior.

**Comment by melanieclarke** [comment]
@emolter - did you have any further comments on this PR?

**Review by emolter — APPROVED** [review]
Looks good to me

**Comment by melanieclarke** [comment]
@tapastro - do you want to review for the maintainers?

**Review by tapastro — APPROVED** [review]

**Pull request merged**
